### ISSUE TITLE
Replace MD5Crypt with Apache Commons Md5Crypt

### ISSUE THREAD

**Issue by ssoloff** [triplea-game/triplea#2901]
Branching a discussion begun in #2764...

The `MD5Crypt` class that was copied into the codebase in the far distant past appears to be licensed such that it is not permissible to use later versions of the GPL.  In addition, the license text in this code simply states "released under the GNU General Public License," so the GPL version isn't even clear; does no version qualifier imply version 1?  IDK.

Regardless, in order to proceed with the upgrade to GPL-3.0, we probably have to remove this code from our source tree.  Apache Commons provides an equivalent implementation in `org.apache.commons.codec.digest.Md5Crypt`.  I ran some tests to see if the two implementations produced identical output.

(Sorry for the length of the following explanation.  Feel free to skip ahead to the TL;DR.)

For passwords that only contain ASCII characters, the implementations produce identical results.  For passwords that contain at least one non-ASCII character (code point > 127), the implementations produce different results.  After looking at the code, my conclusion is that our `MD5Crypt` implementation is at fault for two reasons:

1. It uses the default character set for the platform, and
1. It assumes 1 character is stored in 1 byte.

On my platform the default character set is UTF-8, so that when I use a password with non-ASCII characters, the number of bytes in the encoded array does not equal the number of characters in the string.  This difference causes the algorithms to perform the same crypto operation a different number of times resulting in different output.

The erroneous code in `MD5Crypt#crypt(String, String)` is in two places:

```java
    for (int pl = password.length(); pl > 0; pl -= 16) {
      ctx.update(finalState, 0, pl > 16 ? 16 : pl);
    }
    /*
     * the original code claimed that finalState was being cleared
     * to keep dangerous bits out of memory,
     * but doing this is also required in order to get the right output.
     */
    clearbits(finalState);
    /* Then something really weird... */
    for (int i = password.length(); i != 0; i >>>= 1) {
      if ((i & 1) != 0) {
        ctx.update(finalState, 0, 1);
      } else {
        ctx.update(password.getBytes(), 0, 1);
      }
    }
```

The error in our code is that the `pl` and `i` loop control variables are initialized to `password.length()` instead of `password.getBytes().length`.  Compare the corresponding implementation in `Md5Crypt#md5Crypt(byte[], String, String)`:

```java
        int ii = keyLen;
        while (ii > 0) {
            ctx.update(finalb, 0, ii > 16 ? 16 : ii);
            ii -= 16;
        }

        /*
         * Don't leave anything around in vm they could use.
         */
        Arrays.fill(finalb, (byte) 0);

        /*
         * Then something really weird...
         */
        ii = keyLen;
        final int j = 0;
        while (ii > 0) {
            if ((ii & 1) == 1) {
                ctx.update(finalb[j]);
            } else {
                ctx.update(keyBytes[j]);
            }
            ii >>= 1;
        }
```

where `keyLen` (used to initialize the `ii` loop control variable twice) is equivalent to `password.getBytes().length`.

So, switching to `Md5Crypt` would break lobby logins under the following conditions:

1. User has a lobby password containing at least one non-ASCII character.
1. User is using an old client that only supports MD5Crypt authentication (newer clients favor the bcrypt authenticator, and will bypass the MD5Crypt authenticator).

Note that lobby logins using MD5Crypt are already broken today under the following conditions:

1. User has a lobby password containing at least one non-ASCII character.
1. User is using an old client that only supports MD5Crypt authentication.
1. User is running on a platform whose default character set is not UTF-8.

In this case, the lobby client and server will encode the password strings differently.  I didn't test this, but I suspect a user might not even be able to create an account in this case.

**TL;DR**

* `MD5Crypt` and `Md5Crypt` only produce identical results for ASCII passwords.
* Lobby users would be affected if
    * The lobby is running the new server that uses `Md5Crypt`.
    * The lobby user's password contains at least one non-ASCII character.
    * The lobby user is running a client older than 1.9.0.0.7062 (that doesn't support bcrypt authentication).
* Password-protected network games would be affected if
    * The host is running the new client that uses `Md5Crypt`.
    * The host selects a password containing at least one non-ASCII character.
    * One of the remote players is running a client older than 1.9.0.0.7062 (that doesn't support HMAC-SHA512 authentication).
* Hashed MAC addresses would not be affected because the input string consists solely of ASCII characters.

**Should we replace `MD5Crypt` with Apache Commons `Md5Crypt`?**

**Comment by RoiEXLab** [comment]
@ssoloff
> I didn't test this, but I suspect a user might not even be able to create an account in this case.

If that's the case I would go for it without a doubt.
If it's not we could ask user to install a more recent version of TripleA to get the bcrypt hash added to the DB to fix their account.

That being said we should probably aim to verify emails simply to be able to send password reset emails.

**Comment by ssoloff** [comment]
> I didn't test this, but I suspect a user might not even be able to create an account in this case.

Now that I'm not half-asleep, I researched this a bit more.  The create account scenario I mentioned probably works fine because the hashing _only_ occurs on the lobby client.  The lobby server simply compares the hashed value to what's stored in the DB; it does not recalculate the hashed password itself (because it never receives the plaintext password).

However, what would fail is if the user created an account with a non-ASCII password on a client with one default charset, and then tried to login from a client with a different default charset.  In this case, the bytes that are hashed would be different on the two clients.  I confirmed this by running the following scenario:

1. Create a new account with a non-ASCII password using 3635 on Linux (default charset: UTF-8).
1. Login succeeded with that account using 3635 on Linux.
1. Login failed with that account using 3635 on Windows (default charset: Cp1252).

**Comment by prastle** [comment]
@ssoloff  So in Noobese ;) We need a breaking release?

**Comment by RoiEXLab** [comment]
@prastle Not really as we already changed the system to a completely different one

**Comment by prastle** [comment]
kk

**Comment by ssoloff** [comment]
@prastle As @RoiEXLab said.  Basically, this would only affect users who

1. Use 1.9.0.0.3635 or earlier, and
1. Use a password with at least one non-ASCII character (e.g. `1234àèìòùABCD`).

I suspect there probably aren't a lot of users who fall into category (2).

**Comment by prastle** [comment]
Well if it makes life easier I would love to get them all off of the pre 8000 series
lots of live play headaches seems to occur from multiple engines.

**Comment by ssoloff** [comment]
I prototyped this replacement and discovered two additional functional differences between the implementations related to salt handling:

1. `MD5Crypt` allows an empty salt, while `Md5Crypt` requires a salt with a minimum length of one.  To avoid changing the API, we can silently replace an empty salt with something legal (e.g. just generate a new 8-character salt).  However, if there are existing hashed passwords out there with an empty salt, we would be unable to process them.
1. `MD5Crypt` allows any character in the salt string, while `Md5Crypt` requires all characters to be in the set `[./a-zA-Z0-9]`.  To avoid changing the API, we can silently replace illegal characters with a legal character (e.g. a period `.`).  However, if there are existing hashed passwords out there with an illegal character in the salt, we would be unable to process them.

I did a quick search of the code and couldn't locate any empty salts or salts that weren't generated using `newSalt()` (which uses the character set `[a-zA-Z0-9]`, and is thus legal).  The only hard-coded salt I found was `"MH"` for hashed MACs.

Looking through the lobby database, I couldn't find any MD5-Crypted passwords that had an empty hash.

So, it would appear that these functional differences may not actually affect us.

**Comment by ssoloff** [comment]
@DanVanAtta @RoiEXLab @ron-murhammer I submitted #2909, which captures the essence of this issue.  The PR description summarizes all the functional changes I've noted above, so you shouldn't have to slog through all the comments here.

**Comment by ssoloff** [comment]
@DanVanAtta @ron-murhammer Do either of you have any concerns over the breaking changes in #2909?  This is the last roadblock preventing us from upgrading to GPL-3.0.

**Comment by ron-murhammer** [comment]
@ssoloff Does it make sense to wait on the breaking changes for the next incompatible release then?

**Comment by ssoloff** [comment]
@ron-murhammer The only issue is that we'll still have GPL-2/3-incompatible _source_ in the distribution.  Now, we could always argue that it's been that way since at least 2005, so what's one more release? :smile:

Note again that the breaking changes only affect users with a non-ASCII password running 3635 or earlier.  I looked at the production lobby database and here's the latest stats:

* 3,865 total users
* 1,077 users have never logged in with release 7621 or later (i.e. no bcrypt password)

Of those 1,077 users:

* 526 have logged in since 2017-11-01 (vs. 3,066 total)
* 392 have logged in since 2017-12-01 (vs. 2,314 total)
* 254 have logged in since 2018-01-01 (vs. 1,358 total)
* 51 have logged in since 2018-02-01 (vs. 157 total)

So let's assume that we have ~500 users still running 3635.  The question is how many of them have used a non-ASCII character in their password?  I'm assuming that's going to be a _really_ low number.

If we do make the breaking change, and some subset of users can't login because they did use a non-ASCII character in their password, I guess the easiest fix would be to manually reset their password and instruct them to change it when they next login.  Not sure if that's acceptable to the community, though.

**Comment by prastle** [comment]
Works for me 👍  I am curious if it also breaks down between anon and registered?

**Comment by prastle** [comment]
Sorry that wasn't what I was curious about @ssoloff .  I realize that.  I was curious if your numbers included anon users?  Or do they just mean we have 3865 "Registered", users?  Plus Unknown on anon?

**Comment by ssoloff** [comment]
@prastle Sorry.  Yes, those are only registered users.  The database does not track anonymous user stats.

**Comment by prastle** [comment]
Np that was what I thought then.  Just saying it would be nice to know a rough number of how many actually use the service.

**Comment by ssoloff** [comment]
@prastle Sounds like we need an authentication statistics table in the database. :smile:

**Comment by prastle** [comment]
There ya go again talking CHINESE again 😄 What exactly is that?

**Comment by RoiEXLab** [comment]
@prastle Well a table containing statistics about lobby logins 🤔

**Comment by prastle** [comment]
I am just saying after all these discussions isn't the real ? we should all be truly asking ourselves is how much is the service used?

**Comment by prastle** [comment]
Example 20 a day in the lobby 200?

**Comment by ssoloff** [comment]
@prastle Yeah, for example, something like recording the count of registered and anonymous logins each day.

**Comment by prastle** [comment]
Yes
Even if no statistics it would give us an idea of where to go from here

**Comment by prastle** [comment]
Sorry no Name data

**Comment by ssoloff** [comment]
@prastle I created a new forum topic for this: https://forums.triplea-game.org/topic/554/recording-lobby-statistics

**Comment by ron-murhammer** [comment]
@ssoloff I see. I didn't realize it was only the small subset of users. I'm alright merging it now then and I think this was then that last piece we were waiting on before cutting a final compatible release before working towards the next incompatible release.
